core/modules/map.rs

```diff
@@ -88,6 +88,95 @@ pub(crate) struct ModuleMap {
   pub(crate) dyn_module_evaluate_idle_counter: Cell<u32>,
 }
 
+/// Map of [`ModuleName`] and [`AssertedModuleType`] to a data field.
+struct ModuleNameTypeMap<T> {
```

Comment:
We use a vec of maps rather than a HashMap with multiple keys to simplify lookups here.